database/manager.go

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package database
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"go.uber.org/zap"
+)
+
+//
+// DatabaseManager writes all committed keys to the database on a timer.
+//
+
+type DatabaseManager struct {
+	logger      logging.Logger
+	keyManagers map[RelayerID]*keyManager
+	interval    time.Duration
+	db          RelayerDatabase
+}
+
+func NewDatabaseManager(logger logging.Logger, db RelayerDatabase, interval time.Duration, keys []RelayerID) *DatabaseManager {
+	keyManagers := make(map[RelayerID]*keyManager)
+	for _, key := range keys {
+		keyManager := newKeyManager(logger)
+		go keyManager.run()
+		keyManagers[key] = keyManager
+	}
+	return &DatabaseManager{
+		logger:      logger,
+		db:          db,
+		interval:    interval,
+		keyManagers: keyManagers,
+	}
+}
+
+// Run writes all committed keys to the database on the configured timer.
+// This function should only be called once.
+func (dm *DatabaseManager) Run() {
+	for range time.Tick(dm.interval) {
+		for id, km := range dm.keyManagers {
+			// Ensure we're not writing the default value
+			if km.maxCommittedHeight > 0 {
+				storedHeight, err := getLatestProcessedBlockHeight(dm.db, id)
+				if err != nil && !IsKeyNotFoundError(err) {
+					dm.logger.Error("Failed to get latest processed block height", zap.Error(err))
+					continue
+				}
+				if storedHeight >= km.maxCommittedHeight {
+					continue
+				}
+				dm.logger.Debug(
+					"writing height",
+					zap.Uint64("height", km.maxCommittedHeight),
+					zap.String("relayerID", id.ID.String()),
+				)
+				err = dm.db.Put(id.ID, LatestProcessedBlockKey, []byte(strconv.FormatUint(km.maxCommittedHeight, 10)))
+				if err != nil {
+					dm.logger.Error("Failed to write latest processed block height", zap.Error(err))
```

Comment:
do the key managers have some identifier that can be logged here too?

Reply:
how are you thinking of detecting if one of many key managers are in a stuck state and can't erroring when going to its iterations?